File: core/report/meals_at_creche/meals_at_creche.js
frappe.query_reports["Meals at Creche"] = {
    filters: [
        {
            fieldname: "partner",
            label: __("Partner"),
            fieldtype: "Link",
            options: "Partner",
            default: "",
            ignore_case: true,
        },
        {
            fieldname: "state",
            label: __("State"),
            fieldtype: "Link",
            options: "State",
            default: "",
            ignore_case: true,
        },
        {
            fieldname: "district",
            label: __("District"),
            fieldtype: "Link",
            options: "District",
            default: "",
            ignore_case: true,
        },
        {
            fieldname: "block",
            label: __("Block"),
            fieldtype: "Link",
            options: "Block",
            default: "",
            ignore_case: true,
        },
        {
            fieldname: "gp",
            label: __("Gram Panchayat"),
            fieldtype: "Link",
            options: "Gram Panchayat",
            default: "",
            ignore_case: true,
        },
        {
            fieldname: "village",
            label: __("Village"),
            fieldtype: "Link",
            options: "Village",
            default: "",
            ignore_case: true,
        },
        {
            fieldname: "creche",
            label: __("Creche"),
            fieldtype: "Link",
            options: "Creche",
            default: "",
            ignore_case: true,
        },
        {
            fieldname: "month",
            label: __("Month"),
            fieldtype: "Select",
            options: [
                { "value": "01", "label": "January" },
                { "value": "02", "label": "February" },
                { "value": "03", "label": "March" },
                { "value": "04", "label": "April" },
                { "value": "05", "label": "May" },
                { "value": "06", "label": "June" },
                { "value": "07", "label": "July" },
                { "value": "08", "label": "August" },
                { "value": "09", "label": "September" },
                { "value": "10", "label": "October" },
                { "value": "11", "label": "November" },
                { "value": "12", "label": "December" }
            ],
            default: frappe.datetime.get_today().split('-')[1],
        },
        {
            fieldname: "year",
            label: __("Year"),
            fieldtype: "Select",
            options: [
                "2024", "2025", "2026", "2027", "2028", "2029", "2030",
                "2031", "2032", "2033", "2034", "2035", "2036", "2037",
                "2038", "2039", "2040", "2041", "2042", "2043", "2044",
                "2045", "2046", "2047", "2048", "2049", "2050"
            ],
            default: frappe.datetime.get_today().split('-')[0],
        },
    ],
};

File: core/report/meals_at_creche/meals_at_creche.py
import frappe
from frappe.utils import nowdate

def execute(filters=None):
    columns = get_columns()
    data = get_summary_data(filters)
    return columns, data

def get_columns():
    columns = [
        {"label": "Partner", "fieldname": "partner", "fieldtype": "Data", "width": 120},
        {"label": "State", "fieldname": "state", "fieldtype": "Data", "width": 120},
        {"label": "District", "fieldname": "district", "fieldtype": "Data", "width": 120},
        {"label": "Block", "fieldname": "block", "fieldtype": "Data", "width": 120},
        {"label": "GP", "fieldname": "gp", "fieldtype": "Data", "width": 120},
        {"label": "Village", "fieldname": "village", "fieldtype": "Data", "width": 120},
        {"label": "Creche", "fieldname": "creche", "fieldtype": "Data", "width": 120},
        {"label": "Creche Open", "fieldname": "creche_open", "fieldtype": "Int", "width": 120},
        {"label": "Breakfast Days", "fieldname": "breakfast_days", "fieldtype": "Int", "width": 120},
        {"label": "Lunch Days", "fieldname": "lunch_days", "fieldtype": "Int", "width": 120},
        {"label": "Egg Days", "fieldname": "egg_days", "fieldtype": "Int", "width": 120},
        {"label": "Evening Snacks Days", "fieldname": "evening_snacks_days", "fieldtype": "Int", "width": 120},
    ]
    return columns

@frappe.whitelist()
def get_summary_data(filters=None):
    partner = filters.get("partner") if filters else None
    state = filters.get("state") if filters else None
    district = filters.get("district") if filters else None
    block = filters.get("block") if filters else None
    gp = filters.get("gp") if filters else None
    village = filters.get("village") if filters else None
    creche = filters.get("creche") if filters else None
    month = filters.get("month") if filters else None
    year = filters.get("year") if filters else None

    # Default values if not provided
    if not month:
        month = nowdate().split('-')[1]
    if not year:
        year = nowdate().split('-')[0]

    conditions = []
    if partner:
        conditions.append(f"AND p.partner_name = '{partner}'")
    if state:
        conditions.append(f"AND s.state_name = '{state}'")
    if district:
        conditions.append(f"AND d.district_name = '{district}'")
    if block:
        conditions.append(f"AND b.block_name = '{block}'")
    if gp:
        conditions.append(f"AND g.gp_name = '{gp}'")
    if village:
        conditions.append(f"AND v.village_name = '{village}'")
    if creche:
        conditions.append(f"AND c.creche_name = '{creche}'")

    condition_str = " ".join(conditions)

    # Filtering data by month and year
    date_condition = f"AND MONTH(att.date_of_attendance) = {month} AND YEAR(att.date_of_attendance) = {year}"

    sql_query = f"""
        SELECT
            p.partner_name AS partner,
            s.state_name AS state,
            d.district_name AS district,
            b.block_name AS block,
            g.gp_name AS gp,
            v.village_name AS village,
            c.creche_name AS creche,
            COUNT(DISTINCT CASE WHEN att.date_of_attendance IS NOT NULL THEN att.name END) AS creche_open,
            COUNT(DISTINCT CASE WHEN att.breakfast IS NOT NULL AND att.breakfast > 0 THEN att.date_of_attendance END) AS breakfast_days,
            COUNT(DISTINCT CASE WHEN att.lunch IS NOT NULL AND att.lunch > 0 THEN att.date_of_attendance END) AS lunch_days,
            COUNT(DISTINCT CASE WHEN att.egg IS NOT NULL AND att.egg > 0 THEN att.date_of_attendance END) AS egg_days,
            COUNT(DISTINCT CASE WHEN att.evening_snacks IS NOT NULL AND att.evening_snacks > 0 THEN att.date_of_attendance END) AS evening_snacks_days
        FROM
            `tabChild Attendance` AS att
        JOIN
            `tabCreche` AS c ON att.creche_id = c.name
        JOIN
            `tabState` AS s ON c.state_id = s.name
        JOIN
            `tabDistrict` AS d ON c.district_id = d.name
        JOIN
            `tabBlock` AS b ON c.block_id = b.name
        JOIN
            `tabGram Panchayat` AS g ON c.gp_id = g.name
        JOIN
            `tabVillage` AS v ON c.village_id = v.name
        JOIN
            `tabPartner` AS p ON c.partner_id = p.name
        WHERE
            1=1
            {condition_str}
            {date_condition}
        GROUP BY
            p.partner_name, s.state_name, d.district_name, b.block_name, g.gp_name, v.village_name, c.creche_name
        ORDER BY
            p.partner_name, s.state_name, d.district_name, b.block_name, g.gp_name, v.village_name, c.creche_name
    """

    data = frappe.db.sql(sql_query, as_dict=True)
    return data
